Mage.Sets/src/mage/cards/e/ElkinBottle.java

```diff
@@ -0,0 +1,148 @@
+
+    public ElkinBottle(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId,setInfo,new CardType[]{CardType.ARTIFACT},"{3}");
+
+        // {3}, {tap}, Exile the top card of your library. Until the beginning of your next upkeep, you may play that card.
+        Ability ability = new SimpleActivatedAbility(Zone.BATTLEFIELD, new ElkinBottleExileEffect(), new GenericManaCost(3));
+        ability.addCost(new TapSourceCost());
+        this.addAbility(ability);
+    }
+
+    public ElkinBottle(final ElkinBottle card) {
+        super(card);
+    }
+
+    @Override
+    public ElkinBottle copy() {
+        return new ElkinBottle(this);
+    }
+}
+
+class ElkinBottleExileEffect extends OneShotEffect {
+
+    public ElkinBottleExileEffect() {
+        super(Outcome.Detriment);
+        this.staticText = "Exile the top card of your library. Until the beginning of your next upkeep, you may play that card";
+    }
+
+    public ElkinBottleExileEffect(final ElkinBottleExileEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public ElkinBottleExileEffect copy() {
+        return new ElkinBottleExileEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        Permanent sourcePermanent = game.getPermanentOrLKIBattlefield(source.getSourceId());
+        if (sourcePermanent != null && controller != null && controller.getLibrary().hasCards()) {
+            Library library = controller.getLibrary();
+            Card card = library.removeFromTop(game);
+            if (card != null) {
+                String exileName = new StringBuilder(sourcePermanent.getIdName()).append(" <this card may be played until the beginning of your next upkeep>").toString();
+                controller.moveCardToExileWithInfo(card, source.getSourceId(), exileName, source.getSourceId(), game, Zone.LIBRARY, true);
+                ContinuousEffect effect = new ElkinBottleCastFromExileEffect();
+                effect.setTargetPointer(new FixedTarget(card.getId()));
+                game.addEffect(effect, source);
+            }
+            return true;
+        }
+        return false;
+    }
+}
+
+class ElkinBottleCastFromExileEffect extends AsThoughEffectImpl {
+
+    public ElkinBottleCastFromExileEffect() {
+        super(AsThoughEffectType.PLAY_FROM_NOT_OWN_HAND_ZONE, Duration.Custom, Outcome.Benefit);
+        this.staticText = "Until the beginning of your next upkeep, you may play that card.";
+    }
+    
+    public ElkinBottleCastFromExileEffect(final ElkinBottleCastFromExileEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public ElkinBottleCastFromExileEffect copy() {
+        return new ElkinBottleCastFromExileEffect(this);
+    }
+
+    @Override
+    public boolean isInactive(Ability source, Game game) {
+        if (game.getPhase().getStep().getType() == PhaseStep.UPKEEP) {
```

Comment:
This won't work if activated during your upkeep since it'll immediately become inactive.

Reply:
Still working on this, but searching for a solution, I've noticed the exact same issue in Grinning Totem and Xenic Poltergeist.

Reply:
Yeah, I think the solution would involve some kind of upkeep counter, but it's hard to see exactly how. It's why I keep trying to implement Halfdane and then give up. I may try to add some more cases to the Duration enum since there are a bunch of unimplemented cards with similarly weird durations.

Reply:
Actually it looks like `GainAbilityTargetEffect` has something like that in it, I'm gonna try to extend it to all continuous effects.

Reply:
Alright, got a separate branch open with the changes, also got [confirmation from Eli Shiffrin](https://twitter.com/elishffrn/status/914494507957096448) that Elkin Bottle and Grinning Totem are functionally identical to the other effects that last until upkeep.

Reply:
So do you suggest I should pull out Elkin Bottle until setDurationToPhase gets included into continuous effects?

Reply:
Yeah, probably a good idea.